File: ElementsServices/tests/src/DataSync/DataSyncUtils_test.cpp
/*
 * Copyright (C) 2012-2020 Euclid Science Ground Segment
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 3.0 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include <string>
#include <vector>

#define BOOST_TEST_IGNORE_NON_ZERO_CHILD_CODE
#include <boost/test/unit_test.hpp>

#include "ElementsServices/DataSync/DataSyncUtils.h"

namespace DataSync = ElementsServices::DataSync;

using std::string;
using std::vector;

//-----------------------------------------------------------------------------

BOOST_AUTO_TEST_SUITE(DataSyncUtils_test)

//-----------------------------------------------------------------------------

void checkLower(vector<string> casedList, string uncased) {
  for (const auto& cased : casedList) {
    BOOST_CHECK_EQUAL(DataSync::lower(cased), uncased);
  }
}

BOOST_AUTO_TEST_CASE(lower_test) {
  const vector<string> casedList { "webdav", "WebDAV", "WEBDAV" };
  checkLower(casedList, "webdav");
}

BOOST_AUTO_TEST_CASE(checkCall_test) {
  BOOST_CHECK(DataSync::checkCall("ls"));
  BOOST_CHECK(not DataSync::checkCall("This command should not work"));
}

BOOST_AUTO_TEST_CASE(runCommand_out_test) {
  const string msg("toto");
  const auto outerr = DataSync::runCommandAndCaptureOutErr("echo " + msg);
  auto output = outerr.first;
  const string::size_type size = output.size();
  if (output[size - 1] == '\n') {
    output.resize(size - 1);
  }
  BOOST_CHECK_EQUAL(output, msg);
}

BOOST_AUTO_TEST_CASE(localWorkspacePrefix_test) {
  const string localEv = DataSync::environmentVariable("NOPREFIX");
  if (localEv == "") {
    BOOST_CHECK_NE(DataSync::localWorkspacePrefix(), "");
  } else {
    BOOST_CHECK_EQUAL(DataSync::localWorkspacePrefix(), "");
  }
}

// @TODO runCommand_err_test

BOOST_AUTO_TEST_CASE(containsInThisOrder_test) {

  using DataSync::containsInThisOrder;

  const string input("0 one Two THREE");
  const vector<string> perfect = { "0", "one", "Two", "THREE" };
  const vector<string> partial = { "0", "one", "THR" };
  const vector<string> empty;
  const vector<string> unordered = { "0", "Two", "one", "THREE" };
  const vector<string> random = { "R", "and", "OM" };
  BOOST_CHECK(containsInThisOrder(input, perfect));
  BOOST_CHECK(containsInThisOrder(input, partial));
  BOOST_CHECK(containsInThisOrder(input, empty));
  BOOST_CHECK(not containsInThisOrder(input, unordered));
  BOOST_CHECK(not containsInThisOrder(input, random));
}

//-----------------------------------------------------------------------------

BOOST_AUTO_TEST_SUITE_END()

File: ElementsServices/tests/python/DataSyncUtils_test.py
#
# Copyright (C) 2012-2020 Euclid Science Ground Segment
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3.0 of the License, or (at your option)
# any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#


import os.path
import unittest

from ElementsKernel.Temporary import TempDir, TempEnv
from ElementsServices.DataSync import DataSyncUtils


class TestDataSyncUtils(unittest.TestCase):

    def setUp(self):
        unittest.TestCase.setUp(self)
        self.m_top_dir = TempDir(prefix="DataSync_test")
        self.m_env = TempEnv()
        self.m_env["WORKSPACE"] = os.path.join(self.m_top_dir.path(), "workspace")
        
    def tearDown(self):
        unittest.TestCase.tearDown(self)
        del self.m_top_dir

    def test_configFileFound(self):
        rel_path_to_some_file = "ElementsServices/testdata/test_file_list.txt"
        abs_path_to_some_file = DataSyncUtils.dataSyncConfFilePath(
            rel_path_to_some_file)
        assert os.path.isfile(abs_path_to_some_file)

    def test_runCommandAndCaptureOutErr(self):
        the_message = "Hello, Antoine!"
        out, err = DataSyncUtils.runCommandAndCaptureOutErr(
            "echo " + the_message)
        out = out.replace("\n", "")
        err = err.replace("\n", "")
        assert out == the_message, "Command output misread: " + out + " != " + the_message
        assert not err

    def test_createLocalDirOf(self):
        the_dir = "/tmp/somewhere/some_temporary_test_data"
        the_filename = "some_test_file.txt"
        the_path = DataSyncUtils.concatenatePaths([DataSyncUtils.localWorkspacePrefix(), 
                                                  the_dir, the_filename])
        DataSyncUtils.createLocalDirOf(the_path)
        assert os.path.isdir(DataSyncUtils.concatenatePaths([DataSyncUtils.localWorkspacePrefix(), 
                                                            the_dir]))
        os.rmdir(DataSyncUtils.concatenatePaths([DataSyncUtils.localWorkspacePrefix(), 
                                                the_dir]))

    def test_localWorkspacePrefix(self):
        prefix_ev = DataSyncUtils.environmentVariable('NOPREFIX')
        workspace_ev = DataSyncUtils.environmentVariable('WORKSPACE')
        prefix = DataSyncUtils.localWorkspacePrefix()
        if prefix_ev == '':
            assert prefix == workspace_ev
        else:
            assert prefix == ''

    def checkConcatenatePaths(self, chunks, expected):
        found = DataSyncUtils.concatenatePaths(chunks)
        assert found == expected

    def test_concatenatePaths(self):
        self.checkConcatenatePaths(["", "/tmp"], "/tmp")
        self.checkConcatenatePaths(["", "/tmp/"], "/tmp")
        self.checkConcatenatePaths(["", "/tmp", "file.fits"], "/tmp/file.fits")
        self.checkConcatenatePaths(
            ["", "/tmp/", "/data/file.fits"], "/tmp/data/file.fits")
        self.checkConcatenatePaths(["/prefix", "/tmp"], "/prefix/tmp")
